CLIP

patil-suraj opened the pull request:
# What does this PR do?

This PR adds the [CLIP](https://github.com/openai/CLIP) model.

CLIP is a multi-modal vision+language model which uses a transformer model for encoding both the images and text.

- The model here is designed such that both `CLIPTextModel` and `CLIPVisionModel` can be loaded independently, and composed together to get the `CLIPModel`.
- Both `CLIPTextModel` and `CLIPVisionModel` use the shared encoder class `CLIPEncoder`.
- The config classes are also kept in separate i.e `CLIPTextConfig` and `CLIPVisionConfig`.  This could be kept in one config class but then we would have to add two arguments for each config value i.e `text_hidden_size` for text model `vision_hidden_size` for vision model etc.
One issue here is that when we load an individual model, like `CLIPTextModel` using the weights of the whole `CLIPModel`
the config ends up containing both text and vision config dicts, this does not cause any issue but could be confusing to look at. 

One important thing to note here is that CLIP's tokenizer does have a pad token defined for it, but they use 0 as `pad_token_id`  to pad the text, but the token, but the token associated with 0 is not a pad token. So here, to able to do padding I've added `pad_token_id` as a `property` which returns 0. I would be happy to hear if there is some other way to achieve this.

Also, I've added a processor class here but not sure if we really need it for this model. We could easily use the extractor for the vision model and tokenizer for the text model.

Would love your review about the design @LysandreJik , @patrickvonplaten , @sgugger.

patil-suraj committed:
begin second draft

patil-suraj committed:
fix import, style

patil-suraj committed:
add loss

patil-suraj committed:
fix embeds, logits_scale, and projection

patil-suraj committed:
fix imports

patil-suraj committed:
add conversion script

patil-suraj committed:
add feature_extractor and processor

patil-suraj committed:
style

patil-suraj committed:
add tests for tokenizer, extractor and processor

patil-suraj committed:
add vision model tests

patil-suraj committed:
add weight init

patil-suraj committed:
add more tests

patil-suraj committed:
fix save_load  test

patil-suraj committed:
model output, dosstrings, causal mask

patil-suraj committed:
config doc

patil-suraj committed:
add clip model tests

patil-suraj committed:
return dict

patil-suraj committed:
bigin integration test

patil-suraj committed:
add integration tests

patil-suraj committed:
fix-copies

patil-suraj committed:
fix init

patil-suraj committed:
Clip => CLIP

patil-suraj committed:
fix module name

patil-suraj committed:
docs

patil-suraj committed:
fix doc

patil-suraj committed:
output_dim => projection_dim

patil-suraj committed:
fix checkpoint names

patil-suraj committed:
remoe fast tokenizer file

patil-suraj committed:
fix conversion script

patil-suraj committed:
fix tests, quality

patil-suraj committed:
put causal mask on device

sgugger commented on src/transformers/models/clip/processing_clip.py:
```diff
@@ -0,0 +1,136 @@
+                f"`feature_extractor` has to be of type {CLIPFeatureExtractor.__class__}, but is {type(feature_extractor)}"
+            )
+        if not isinstance(tokenizer, CLIPTokenizer):
+            raise ValueError(f"`tokenizer` has to be of type {CLIPTokenizer.__class__}, but is {type(tokenizer)}")
```
```suggestion
            raise ValueError(f"`tokenizer` has to be of type CLIPTokenizer, but is {type(tokenizer)}")
```

sgugger commented on tests/test_modeling_clip.py:
```diff
@@ -0,0 +1,528 @@
+
+    def prepare_config_and_inputs_for_common(self):
+        config_and_inputs = self.prepare_config_and_inputs()
+        (
+            config,
+            input_ids,
+            input_mask,
+        ) = config_and_inputs
```
```suggestion
        config, input_ids, input_mask = config_and_inputs
```

sgugger commented on tests/test_modeling_clip.py:
```diff
@@ -0,0 +1,528 @@
+
+    def prepare_config_and_inputs_for_common(self):
+        config_and_inputs = self.prepare_config_and_inputs()
+        (
+            config,
+            input_ids,
+            attention_mask,
+            pixel_values,
+        ) = config_and_inputs
```
```suggestion
        config, input_ids, attention_mask, pixel_values = config_and_inputs
```

sgugger commented on tests/test_modeling_clip.py:
```diff
@@ -0,0 +1,528 @@
+
+        image = prepare_img()
+        vision_inputs = feature_extractor(images=image, return_tensors="pt").to(torch_device)
+        text_inputs = tokenizer(["a photo of a cat", "a photo of a dog"], padding=True, return_tensors="pt").to(
+            torch_device
+        )
```
Use two lines:
```suggestion
        text_inputs = tokenizer(["a photo of a cat", "a photo of a dog"], padding=True, return_tensors="pt")
        text_inputs = text_inputs.to(torch_device)
```

sgugger commented on tests/test_processor_clip.py:
```diff
@@ -0,0 +1,176 @@
+    def setUp(self):
+        self.tmpdirname = tempfile.mkdtemp()
+
+        vocab = [
+            "l",
+            "o",
+            "w",
+            "e",
+            "r",
+            "s",
+            "t",
+            "i",
+            "d",
+            "n",
+            "lo",
+            "low</w>",
+            "er</w>",
+            "lowest</w>",
+            "newer</w>",
+            "wider",
+            "<unk>",
+            "<|endoftext|>",
+        ]
```
One one line with style deactivated would be better.

sgugger commented on tests/test_tokenization_clip.py:
```diff
@@ -0,0 +1,167 @@
+        super().setUp()
+
+        # Adapted from Sennrich et al. 2015 and https://github.com/rsennrich/subword-nmt
+        vocab = [
+            "l",
+            "o",
+            "w",
+            "e",
+            "r",
+            "s",
+            "t",
+            "i",
+            "d",
+            "n",
+            "lo",
+            "low</w>",
+            "er</w>",
+            "lowest</w>",
+            "newer</w>",
+            "wider",
+            "<unk>",
+            "<|endoftext|>",
+        ]
```
One one line too

sgugger reviewed (approved):
Great job! I have two main questions on the API:
- is it possible to add a fast version of the tokenizer?
- why are there two classes for the vision model and the text model? They have the exact same forward, so we should only have one class IMO.

patil-suraj commented:
@sgugger 

> is it possible to add a fast version of the tokenizer?

Yes, will add the fast version as well.


> why are there two classes for the vision model and the text model? They have the exact same forward, so we should only have one class IMO.

Added two versions so that one could just load `CLIPTextModel` or `CLIPVisionModle` directly from `CLIPModel`'s weights. If we just keep single modules then it's not possible to load the weights from the full model because then the keys won't match. The `CLIPModel` has these extra keys `text_model` and `vision_model`, hence the two extra modules with the same keys.

This would allow one to use the vision model in some other downstream tasks like adding a liner layer on top or using it as an image encoder in some other settings. Not sure if users will actually want this, but this does not really add much complexity to the code IMO.

patil-suraj commented on src/transformers/models/clip/feature_extraction_clip.py:
```diff
@@ -0,0 +1,156 @@
+logger = logging.get_logger(__name__)
+
+
+class CLIPFeatureExtractor(FeatureExtractionMixin, ImageFeatureExtractionMixin):
```
Sylvain must have had good reason not to inherit from `FeatureExtractionMixin` 😉

patil-suraj commented on src/transformers/models/clip/feature_extraction_clip.py:
```diff
@@ -0,0 +1,156 @@
+    def __call__(
+        self,
+        images: Union[
+            Image.Image, np.ndarray, "torch.Tensor", List[Image.Image], List[np.ndarray], List["torch.Tensor"]  # noqa
```
Yes, that would be better. We could do it in a follow-up PR, this is used in other feature extractors as well (ViT, DeiT)

patil-suraj committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

LysandreJik commented on src/transformers/models/clip/configuration_clip.py:
```diff
@@ -0,0 +1,267 @@
+    """
+    model_type = "clip_text_model"
+
+    def __init__(
```
(nit) The parameters don't align with the docstring

LysandreJik commented on src/transformers/models/clip/configuration_clip.py:
```diff
@@ -0,0 +1,267 @@
+
+    model_type = "clip_vision_model"
+
+    def __init__(
```
(nit) The parameters don't align with the docstring

patil-suraj committed:
fix attribute test

patil-suraj committed:
style

LysandreJik commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,947 @@
+        assert (
+            self.head_dim * self.num_heads == self.embed_dim
+        ), f"embed_dim must be divisible by num_heads (got `embed_dim`: {self.embed_dim} and `num_heads`: {self.num_heads})."
+        self.scaling = self.head_dim ** -0.5
```
Why not `scale` like in the Embeddings?

LysandreJik commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,947 @@
+    ):
+        r"""
+        Args:
+            input_ids (:obj:`torch.LongTensor` of shape :obj:`(batch_size, sequence_length)`):
```
The docstrings are not coherent with the `forward` method

LysandreJik commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,947 @@
+        return_dict=None,
+    ):
+        r"""
+        Args:
```
The `position_ids` are missing from the docstring

LysandreJik commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,947 @@
+        return_dict=None,
+    ):
+        r"""
+        Args:
```
`attention_mask` missing from the docstring

LysandreJik commented on src/transformers/models/clip/tokenization_clip.py:
```diff
@@ -0,0 +1,371 @@
+# coding=utf-8
+# Copyright The HuggingFace Team and The HuggingFace Inc. team. All rights reserved.
```
Boom, 🤗🤗🤗🤗 everywhere

LysandreJik reviewed (approved):
Great work on this PR @patil-suraj! Wonderful to welcome CLIP in the library!

patil-suraj commented on src/transformers/models/clip/configuration_clip.py:
```diff
@@ -0,0 +1,267 @@
+logger = logging.get_logger(__name__)
+
+CLIP_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "valhalla/clip-vit-base-patch32": "https://huggingface.co/valhalla/clip-vit-base-patch32/resolve/main/config.json",
```
yes!

sgugger replied:
Still standing :-)

patil-suraj committed:
address sylvains comments

patil-suraj committed:
style

patil-suraj committed:
fix docstrings

patrickvonplaten commented on docs/source/model_doc/clip.rst:
```diff
@@ -0,0 +1,151 @@
+
+CLIP is a multi-modal vision and language model. It can be used for image-text similarity and for zero-shot image
+classification. CLIP uses a ViT like transformer to get visual features and a causal language model to get the text
+features. Both the text and visual features are then projected to a latent space with similar dimentions. The dot
```
(nit) think it's identical dimensions no? (also small typo in dimensions)

patrickvonplaten commented on docs/source/model_doc/clip.rst:
```diff
@@ -0,0 +1,151 @@
+        >>> from transformers import CLIPFeatureExtractor, CLIPTokenizer, CLIPModel
+
+        >>> model = CLIPModel.from_pretrained("./hf_model_v2")
+        >>> tokenizer = CLIPTokenizer.from_pretrained("./hf_model_v2/")
```
doesn't make sense yet (or at all?) to add a processor here for both image & text?

patrickvonplaten replied:
seems there is a processor -> should this not be used here for the easiest code example?

patrickvonplaten commented on src/transformers/models/clip/configuration_clip.py:
```diff
@@ -0,0 +1,266 @@
+            Number of attention heads for each attention layer in the Transformer encoder.
+        image_size (:obj:`int`, `optional`, defaults to :obj:`224`):
+            The size (resolution) of each image.
+        patch_size (:obj:`int`, `optional`, defaults to :obj:`32`):
```
```suggestion
        patch_size (:obj:`int`, `optional`, defaults to 32):
```

patrickvonplaten commented on src/transformers/models/clip/configuration_clip.py:
```diff
@@ -0,0 +1,266 @@
+            Number of hidden layers in the Transformer encoder.
+        num_attention_heads (:obj:`int`, `optional`, defaults to 12):
+            Number of attention heads for each attention layer in the Transformer encoder.
+        image_size (:obj:`int`, `optional`, defaults to :obj:`224`):
```
```suggestion
        image_size (:obj:`int`, `optional`, defaults to 224):
```

patrickvonplaten commented on src/transformers/models/clip/configuration_clip.py:
```diff
@@ -0,0 +1,266 @@
+        >>> from transformers import CLIPVisionModel, CLIPTextConfig
+
+        >>> # Initializing a CLIPVisionModel with valhalla/clip-vit-base-patch32 style configuration
+        >>> configuration = CLIPConfig()
```
```suggestion
        >>> configuration = CLIPVisionConfig()
```

patrickvonplaten commented on src/transformers/models/clip/configuration_clip.py:
```diff
@@ -0,0 +1,266 @@
+
+        Example::
+
+        >>> from transformers import CLIPVisionModel, CLIPTextConfig
```
```suggestion
        >>> from transformers import CLIPVisionModel, CLIPVisionConfig
```

patrickvonplaten commented on src/transformers/models/clip/configuration_clip.py:
```diff
@@ -0,0 +1,266 @@
+
+        Example::
+
+        >>> from transformers import CLIPTextModel, CLIPTextConfig
```
```suggestion
        >>> from transformers import CLIPTextModel, CLIPTextConfig
```

patrickvonplaten commented on src/transformers/models/clip/configuration_clip.py:
```diff
@@ -0,0 +1,266 @@
+    model_type = "clip"
+    is_composition = True
+
+    def __init__(self, projection_dim=512, **kwargs):
```
```suggestion
    def __init__(self, text_config_dict, vision_config_dict, projection_dim=512, **kwargs):
```

IMO, it should not be possible to instantiate a `CLIPConfig` without passing the two configs

patrickvonplaten replied:
or add an assert that both configs have to be in kwargs

patrickvonplaten commented on src/transformers/models/clip/feature_extraction_clip.py:
```diff
@@ -0,0 +1,156 @@
+            Whether to resize the input to a certain :obj:`size`.
+        size (:obj:`int`, `optional`, defaults to 224):
+            Resize the input to the given size. Only has an effect if :obj:`do_resize` is set to :obj:`True`.
+        resample (:obj:`int`, `optional`, defaults to :obj:`PIL.Image.BICUBIC`):
```
would advocate for the same `resampling_filter` 

`resample` sounds like a boolean flag to me

patil-suraj replied:
This argument name is used in other extractors as well (ViT, DeiT) so let's change this in a follow-up PR,

patrickvonplaten commented on src/transformers/models/clip/feature_extraction_clip.py:
```diff
@@ -0,0 +1,156 @@
+            :obj:`True`.
+        do_normalize (:obj:`bool`, `optional`, defaults to :obj:`True`):
+            Whether or not to normalize the input with :obj:`image_mean` and :obj:`image_std`.
+        image_mean (:obj:`List[int]`, defaults to :obj:`[0.485, 0.456, 0.406]`):
```
just out of curiosity - where do those numbers come from? :D Is that the standard image mean?

patil-suraj replied:
these are from official CLIP repo

https://github.com/openai/CLIP/blob/cfcffb90e69f37bf2ff1e988237a0fbe41f33c04/clip/clip.py#L64

patrickvonplaten commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,938 @@
+]
+
+
+def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
```
add # Copied from here maybe (feels like I've seen this before :-))

patrickvonplaten commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,938 @@
+
+        scale = self.embed_dim ** -0.5
+        class_embed_weights = torch.empty(self.embed_dim)
+        nn.init.normal_(class_embed_weights, mean=0.0, std=scale)
```
why not put in `_init_weights(...)`?

patrickvonplaten commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,938 @@
+        self.num_patches = (self.image_size // self.patch_size) ** 2
+        self.num_positions = self.num_patches + 1
+        self.position_embedding = nn.Embedding(self.num_positions, self.embed_dim)
+        self.register_buffer("position_ids", torch.arange(self.num_positions).expand((1, -1)))
```
not sure - isn't `torch.device` needed in the `torch.arange(...)` ?

patil-suraj replied:
No, we don't need it. `position_ids` will be put on the same device when we call `.to(device)` on the moduel

patrickvonplaten commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,938 @@
+        self.position_embedding = nn.Embedding(config.max_position_embeddings, embed_dim)
+
+        # position_ids (1, len position emb) is contiguous in memory and exported when serialized
+        self.register_buffer("position_ids", torch.arange(config.max_position_embeddings).expand((1, -1)))
```
same about torch_device here

patrickvonplaten commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,938 @@
+        else:
+            input_shape = inputs_embeds.size()[:-1]
+
+        seq_length = input_shape[1]
```
maybe nicer to save some lines

`seq_length = input_ids.shape[-1] if input_ids is not None else inputs_embeds.shape[-2]`

patrickvonplaten commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,938 @@
+class CLIPMLP(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
```
```suggestion
```

patil-suraj replied:
This is a shared module between vision and text model, so we need to save the config here,  which is then used in `_init_weights`

patrickvonplaten commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,938 @@
+class CLIPTextEmbeddings(nn.Module):
+    def __init__(self, config: CLIPTextConfig):
+        super().__init__()
+        self.config = config
```
```suggestion
        self.config = config
```

patil-suraj replied:
Same as above

patrickvonplaten commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,938 @@
+
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
```
```suggestion
        self.config = config
```

patil-suraj replied:
same as above

patrickvonplaten commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,938 @@
+
+    def __init__(self, config: CLIPConfig):
+        super().__init__()
+        self.config = config
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,952 @@
+
+        Returns:
+        """
+        return self.text_model(
```
100% agree with @sgugger here, was having the same question

patrickvonplaten commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,938 @@
+    def __init__(self, config: CLIPConfig):
+        super().__init__(config)
+
+        text_config = config.text_config
```
think it would make sense here to check that config.text_config is of type `CLIPTextConfig` and raise a nice error if not

sgugger replied:
Still standing!

patrickvonplaten commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,938 @@
+        output_attentions=None,
+        output_hidden_states=None,
+        return_dict=None,
+    ):
```
docstring would be coool

patrickvonplaten commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,938 @@
+
+        return text_embeds
+
+    def encode_image(
```
docstring would be nice, but like it very much to provide these functions

patrickvonplaten replied:
Actually, thinking a bit more about it isn't it a bit weird that `encode_image(...)` does not return the `last_hidden_states`? When I think about "encoding" I would expect the encoded hidden_states to be returned (maybe return both?!)

patrickvonplaten replied:
Or maybe give it a better (different) name?

patrickvonplaten commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -0,0 +1,938 @@
+
+        """
+        return_dict = return_dict if return_dict is not None else self.config.return_dict
+        vision_outputs = self.vision_model(
```
why not call the above functions here?!

patrickvonplaten commented on src/transformers/models/clip/processing_clip.py:
```diff
@@ -0,0 +1,136 @@
+                  :meth:`~transformers.PreTrainedFeatureExtractor.save_pretrained` method, e.g.,
+                  ``./my_model_directory/``.
+                - a path or url to a saved feature extractor JSON `file`, e.g.,
+                  ``./my_model_directory/feature_extraction_config.json``.
```
```suggestion
                  ``./my_model_directory/preprocessor_config.json``.
```

patrickvonplaten commented on src/transformers/models/clip/processing_clip.py:
```diff
@@ -0,0 +1,136 @@
+                This can be either:
+
+                - a string, the `model id` of a pretrained feature_extractor hosted inside a model repo on
+                  huggingface.co. Valid model ids can be located at the root-level, like ``bert-base-uncased``, or
```
maybe use clip checkpoint instead of bert here

patrickvonplaten commented on src/transformers/models/clip/tokenization_clip.py:
```diff
@@ -0,0 +1,371 @@
+            re.IGNORECASE,
+        )
+
+    # Very ugly hack to enable padding
```
no way to make this cleaner?

patrickvonplaten commented on tests/test_modeling_clip.py:
```diff
@@ -0,0 +1,522 @@
+            intermediate_size=self.intermediate_size,
+            dropout=self.dropout,
+            attention_dropout=self.attention_dropout,
+            is_decoder=False,
```
can the model be a decoder?

patrickvonplaten commented on tests/test_modeling_clip.py:
```diff
@@ -0,0 +1,522 @@
+        self.model_tester.create_and_check_model(*config_and_inputs)
+
+    def test_attention_outputs(self):
+        config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
```
this test can be cleaned up, no? we don't need all those `chunk_length`, `key-length`, `encoder_seq_length`, ...

patrickvonplaten commented on tests/test_modeling_clip.py:
```diff
@@ -0,0 +1,522 @@
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_model(*config_and_inputs)
+
+    def test_training(self):
```
Can we test training somehow? Is it tested at all?

patrickvonplaten reviewed (commented):
Very cool! Left a couple of comments.

In general, I very much like the design. A couple of things to improve IMO:
- as @sgugger mentioned, I don't see why we have a `CLIPTextModel` and a `CLIPTextTransformer` (same for vision)
- think the methods `encode_image` and `encode_text` should be rethought a bit (I would have expected the fn to return the `last_hidden_states`) -> maybe just renaming the function is enough or a clear docstring
- Do we have tests for training?

patil-suraj committed:
add qucik_gelu in activations, docstrings

patil-suraj committed:
clean-up attention test

patil-suraj committed:
fix act fun

patil-suraj committed:
fix config

patil-suraj committed:
fix torchscript tests

patil-suraj committed:
even batch_size

patil-suraj committed:
remove comment

patil-suraj committed:
Merge branch 'master' into add-clip

patil-suraj committed:
fix ouput tu_tuple

patil-suraj committed:
fix save load tests

patil-suraj committed:
fix add tokens test

patil-suraj committed:
add fast tokenizer

patil-suraj committed:
update copyright

patil-suraj committed:
new processor API

patil-suraj committed:
fix docs

patil-suraj committed:
docstrings

patil-suraj committed:
docs

patil-suraj committed:
fix doc

patil-suraj committed:
fix doc

patil-suraj committed:
fix tokenizer

patil-suraj committed:
fix import in doc example

patil-suraj commented:
All green!! 
I've addressed most of the suggestions, notably

- new processor API => as discussed with @patrickvonplaten and @LysandreJik processor's `__call__` now accepts both the text and/or images and returns a single encoding dict. `as_target_processor` is now removed. The API is as follows

```python3
model = CLIPModel.from_pretrained(checkpoint)
inputs = CLIPProcessor(texts=..., images=..., some_other_kwargs)
outputs = model(**inputs)
```
- the `encode_text` and `encode_image` methods are renamed to `get_text_features` and `get_image_features`
- Added fast tokenizer.

Ready for second review @LysandreJik @sgugger @patrickvonplaten

sgugger commented on docs/source/model_doc/clip.rst:
```diff
@@ -0,0 +1,157 @@
+        >>> model = CLIPModel.from_pretrained("openai/clip-vit-base-patch32")
+        >>> processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")
+
+        >>> def prepare_img():
+        ...    url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+        ...    im = Image.open(requests.get(url, stream=True).raw)
+        ...    return im
+        >>> image = prepare_img()
```
Given the fact we are only using this function once, I would just replace this by
```suggestion
        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
        >>> image = Image.open(requests.get(url, stream=True).raw)
```

sgugger commented on src/transformers/models/clip/__init__.py:
```diff
@@ -0,0 +1,82 @@
+# There's no way to ignore "F401 '...' imported but unused" warnings in this
+# module, but to preserve other warnings. So, don't check this module at all.
+
+# Copyright 2020 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2021 The HuggingFace Team. All rights reserved.
```

sgugger commented on src/transformers/models/clip/processing_clip.py:
```diff
@@ -0,0 +1,170 @@
+    :meth:`~transformers.CLIPProcessor.decode` for more information.
+
+    Args:
+        feature_extractor (:obj:`CLIPFeatureExtractor`):
```
```suggestion
        feature_extractor (:class:`~transformers.CLIPFeatureExtractor`):
```

sgugger commented on src/transformers/models/clip/processing_clip.py:
```diff
@@ -0,0 +1,170 @@
+    Args:
+        feature_extractor (:obj:`CLIPFeatureExtractor`):
+            An instance of :class:`~transformers.CLIPFeatureExtractor`. The feature extractor is a required input.
+        tokenizer (:obj:`CLIPTokenizer`):
```
```suggestion
        tokenizer (:class:`~transformers.CLIPTokenizer`):
```

sgugger commented on src/transformers/models/clip/processing_clip.py:
```diff
@@ -0,0 +1,170 @@
+
+    Args:
+        feature_extractor (:obj:`CLIPFeatureExtractor`):
+            An instance of :class:`~transformers.CLIPFeatureExtractor`. The feature extractor is a required input.
```
```suggestion
            The feature extractor is a required input.
```
You said it's an instance of that on the previous line ;-)

sgugger commented on src/transformers/models/clip/processing_clip.py:
```diff
@@ -0,0 +1,170 @@
+        feature_extractor (:obj:`CLIPFeatureExtractor`):
+            An instance of :class:`~transformers.CLIPFeatureExtractor`. The feature extractor is a required input.
+        tokenizer (:obj:`CLIPTokenizer`):
+            An instance of :class:`~transformers.CLIPTokenizer`. The tokenizer is a required input.
```
```suggestion
            The tokenizer is a required input.
```
Same here.

sgugger commented on src/transformers/models/clip/processing_clip.py:
```diff
@@ -0,0 +1,170 @@
+                  :meth:`~transformers.PreTrainedFeatureExtractor.save_pretrained` method, e.g.,
+                  ``./my_model_directory/``.
+                - a path or url to a saved feature extractor JSON `file`, e.g.,
+                  ``./my_model_directory/preprocessor_config.json``.
+            **kwargs
```
```suggestion
                  ``./my_model_directory/preprocessor_config.json``.

            **kwargs
```
Probably needed for sphinx to be happy.

sgugger reviewed (approved):
Great work! A few last loose ends to tie up (in particular don't forget to replace all checkpoint names in the docstrings by ones in the openai namespace!) and it should be good to merge.

patil-suraj committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

patil-suraj committed:
check types of config

LysandreJik reviewed (approved):
Great, LGTM! Only need to update from `valhalla` namespace to `openai` namespace and it looks good to me.

patil-suraj committed:
valhalla => openai

patil-suraj committed:
load image using url

patil-suraj committed:
fix test

patil-suraj committed:
typo

lycfight commented:
> Hi @lycfight could you please open an issue with a minimal code snippet so we could take a look. Thanks :)

of course
